### ISSUE TITLE
Fun intersection with apc and Macs running Vagrant

### ISSUE THREAD

**Issue by muglug** [composer/xdebug-handler#57]
When XDebug is turned on, and the XDebugHandler is initialised, on a Vagrant boxes with a CentOS installation, there's the following error when running Psalm:

```
PHP Fatal error:  PHP Startup: apc_mmap: mkstemp on /tmp/apc.KyxDdB failed: in Unknown on line 0
```

It's triggered when calling `passthru`.

This *only* happens when running Vagrant on macs, not on Linux.

I very much doubt this is the *fault* of XDebugHandler, but it obviously triggers the issue.

**Comment by johnstevenson** [comment]
Does just running `passthru` in a simple script (calling itself) cause this issue? I could always add `apc.enable_cli=0` to the tmp ini (as we do for opcache) but it would be good to know what is actually causing this first.

**Comment by johnstevenson** [comment]
Your script isn't calling itself (the idea is to mimic xdebug-handler).

**Comment by muglug** [comment]
ah sorry
```php
<?php
if (rand(0, 1)) {
  echo 'hello' . PHP_EOL;
  passthru('php ' . __FILE__);
}
``

that works fine.

Would it be possible to pass in a list of extensions to disable? We have an unrelated issue due to https://github.com/grpc/grpc/issues/13412 where we want to run Psalm in an environment where that extension is installed. It would be wonderful to use the work you've done here to disable that extension just while Psalm is running.

**Comment by johnstevenson** [comment]
Thanks for the info. So it must be something that xdebug-handler is doing. If you could show me your Vagrant file (or tell me your setup, including the apc ini settings) I'll see if I can reproduce this on Windows, for what its worth.

> Would it be possible to pass in a list of extensions to disable?

As of 1.1.0 you can extend the class to do this sort of stuff. For example, presuming you want to restart if grpc is loaded even if xdebug is not:

```php
use Composer\XdebugHandler\XdebugHandler;

class PsalmRestarter extends XdebugHandler
{
    private $required;

    protected function requiresRestart($loaded)
    {
        $this->required = extension_loaded('grpc');
        return $loaded || $this->required;
    }

    protected function restart($command)
    {
        if ($this->required) {
            $regex = '/^\s*(extension\s*=.*grpc.*)$/mi';
            $content = file_get_contents($this->tmpIni);

            $content = preg_replace($regex, ';$1', $content);
            file_put_contents($this->tmpIni, $content);
        }

        parent::restart($command);
    }
}
```

It is all documented in the readme.

**Comment by johnstevenson** [comment]
Closing - no further info.

**Comment by muglug** [comment]
Thanks for this. I ended up using the code above, but adding the ability to specify extensions: https://github.com/vimeo/psalm/blob/master/src/Psalm/Fork/PsalmRestarter.php

That enabled `apc` to be disabled via the command-line when running inside vagrant on macs. It's a really minor issue though, because we generally run PHP command line utils outside Vagrant.

**Comment by johnstevenson** [comment]
@muglug The apc issue has been fixed in [1.2.1](https://github.com/composer/xdebug-handler/releases/tag/1.2.1)

**Comment by muglug** [comment]
Awesome, thanks!
